Directory *imports* throwing in ESM node

**How to reproduce:**

1. Add an entrypoint: `"preconstruct": { "entrypoints": [ "index.ts", "foo.ts"] }`
2. packages a library, so the entrypoint becomes a "proxy" directory with `package.json` inside it
3. attempt to import that from node using native ESM implementation there: `import { foo } from "pkg/foo";`
4. end up with an error:
> Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import '/some-app/node_modules/pkg/foo' is not supported resolving ES modules imported from /some-app/index.js

**Repro:**

https://github.com/markdalgleish/preconstruct-node-esm-import-issue/tree/f181ab2243c98255723ee106ae665df1fc590db6

**Possible solution:**

Emit `package.json#exports` with the appropriate fields for the declared entrypoints. If we do this just for CJS, without attempting to support ESM bundles in node then it should work and not break anything (like with everything related to this... we should still be extra careful about this though). 

This would limit what can be imported/required in node from a package. However, the whole point of Preconstruct is to only allow what is declared so if somebody has been reaching into "internal" files that's kinda on them. This could be added under an experimental flag if we are worried about the unwanted side-effects. Optionally we could consider always adding `package.json` to the exports as that seems to be a pretty common requirement and I don't see why somebody would really want to "hide" this file.

cc @markdalgleish

> what kind of breaking changes this can introduce if this doesnt try to introduce ESM support in node?

The general "i was importing `x`, `x` isn't in the `exports` field" which of course is a matter of interpretation as to whether importing some file that isn't in the added `exports` field is a breaking change or not since if it's not in the `exports` field, it was probably never intended to be public and some may see it a breaking change and some may not. I don't think Preconstruct should potentially force a package author to do a major if they think adding an `exports` field would be major. This is similar to how the dist file names have a way to go back to the old names if a package author thinks it would be breaking to change it and doesn't want to do a major. 

> What is emitted for that today? Maybe type-level import() wont be limited by exports the same way as runtime is? If this would fail is gonna become a real PITA for users as we dont usually think about if the inferred type can be emitted based on public exports of other pkg. Either way - this seems to be a TS issue and not a Preconstruct one.

On 4.5, you would get a import to `pkg-a/src/a`. (or similar but pointing to declarations)

On the latest nightly as of this comment (`4.6.0-dev.20211126`), you get something like this:

```ts
src/index.ts:2:14 - error TS2742: The inferred type of 'a' cannot be named without a reference to '../node_modules/blah/src/a'. This is likely not portable. A type annotation is necessary.

2 export const a = x();
               ~
```

(I didn't test exactly the example shown above but the same general thing)

To be clear, I don't think this is something that should be "fixed". The current status quo of packages easily having implicit dependencies on their dependencies internal module structure imo is a worse problem than breaking once(or not at all if you don't need to generate declarations that need to reference something that isn't exported). Getting an error upfront and being able to address it is much better imo than some package changing their internal structure and your package that depends on it breaking.

> Agreed, but i think that package.json should always be exported.

Yeah

> I would skip that for now - it sounds niche, potentially YAGNI

I agree that it can be skipped it for now but I don't think it's that niche, having some other js/json file or etc. that you want to allow people to import from and isn't managed by Preconstruct seems like a reasonably common use case. (Emotion does that for example)

> src/index.ts:2:14 - error TS2742: The inferred type of 'a' cannot be named without a reference to '../node_modules/blah/src/a'. This is likely not portable. A type annotation is necessary.

Eh, that could really use a suggestion from TS on how to fix it 🙄 I expect many people to trip over this.

> To be clear, I don't think this is something that should be "fixed". 

Yeah, I got your intention - was just wondering what are the exact implications. Nice catch on this one.

> I agree that it can be skipped it for now but I don't think it's that niche, having some other js/json file or etc. that you want to allow people to import from and isn't managed by Preconstruct seems like a reasonably common use case. (Emotion does that for example)

IIRC Emotion only "exports" macros this way - and those entrypoints could be easily included in `./src`

--- a/.changeset/many-socks-camp.md
+++ b/.changeset/many-socks-camp.md
@@ -1,0 +1,5 @@
+---
+"@preconstruct/cli": minor
+---
+
+[Support for the exports field](https://preconstruct.tools/configuration#exports) has been stabilised. If you were using the experimental flag, you should remove it.

--- a/package.json
+++ b/package.json
@@ -2,6 +2,7 @@
   "name": "@preconstruct/monorepo",
   "version": "0.0.65",
   "description": "A smart build tool for libraries",
+  "packageManager": "yarn@1.22.19",
   "scripts": {
     "build": "yarn preconstruct build",
     "test": "jest",

--- a/packages/cli/src/__tests__/dev.ts
+++ b/packages/cli/src/__tests__/dev.ts
@@ -327,9 +327,6 @@ test("exports field with worker condition", async () => {
           exports: {
             envConditions: ["worker"],
           },
-          ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-            exports: true,
-          },
         },
       }),
       "src/index.js": "console.log(1)",

--- a/packages/cli/src/__tests__/fix.ts
+++ b/packages/cli/src/__tests__/fix.ts
@@ -81,9 +81,6 @@ test("set exports field when opt-in", async () => {
         exports: {
           envConditions: ["worker", "browser"],
         },
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -117,9 +114,6 @@ test("set exports field when opt-in", async () => {
             "worker",
             "browser"
           ]
-        },
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
         }
       }
     }
@@ -137,9 +131,6 @@ test("set exports field when opt-in", async () => {
         exports: {
           envConditions: ["worker", "browser"],
         },
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -173,9 +164,6 @@ test("set exports field when opt-in", async () => {
             "worker",
             "browser"
           ]
-        },
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
         }
       }
     }
@@ -191,9 +179,6 @@ test("set exports field when opt-in with no env conditions", async () => {
       module: "dist/package-exports.esm.js",
       preconstruct: {
         exports: true,
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -215,10 +200,7 @@ test("set exports field when opt-in with no env conditions", async () => {
         "./package.json": "./package.json"
       },
       "preconstruct": {
-        "exports": true,
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
-        }
+        "exports": true
       }
     }
 
@@ -235,9 +217,6 @@ test("set exports field with multiple entrypoints", async () => {
         entrypoints: ["index.js", "other.js", "deep/something.js"],
         exports: {
           envConditions: ["worker", "browser"],
-        },
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
         },
       },
     }),
@@ -315,9 +294,6 @@ test("set exports field with multiple entrypoints", async () => {
             "worker",
             "browser"
           ]
-        },
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
         }
       }
     }
@@ -335,9 +311,6 @@ test("set exports field without root entrypoint", async () => {
             entrypoints: ["other.js"],
             exports: {
               envConditions: ["worker", "browser"],
-            },
-            ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-              exports: true,
             },
           },
         },
@@ -372,9 +345,6 @@ test("set exports field without root entrypoint", async () => {
             "worker",
             "browser"
           ]
-        },
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
         }
       },
       "exports": {
@@ -751,9 +721,6 @@ test("no module field with exports field", async () => {
       main: "dist/pkg-a.cjs.js",
       preconstruct: {
         exports: true,
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -773,10 +740,7 @@ test("no module field with exports field", async () => {
         "./package.json": "./package.json"
       },
       "preconstruct": {
-        "exports": true,
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
-        }
+        "exports": true
       }
     }
 
@@ -795,9 +759,6 @@ test("has browser field but no browser condition", async () => {
       },
       preconstruct: {
         exports: true,
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -827,9 +788,6 @@ test("has browser field but no browser condition", async () => {
           "envConditions": [
             "browser"
           ]
-        },
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
         }
       }
     }
@@ -847,9 +805,6 @@ test("has browser condition but no browser field", async () => {
         exports: {
           envConditions: ["browser"],
         },
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -880,9 +835,6 @@ test("has browser condition but no browser field", async () => {
           "envConditions": [
             "browser"
           ]
-        },
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
         }
       }
     }
@@ -898,9 +850,6 @@ test("preconstruct.exports: true no exports field", async () => {
       module: "dist/pkg-a.esm.js",
       preconstruct: {
         exports: true,
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -920,10 +869,7 @@ test("preconstruct.exports: true no exports field", async () => {
         "./package.json": "./package.json"
       },
       "preconstruct": {
-        "exports": true,
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
-        }
+        "exports": true
       }
     }
 
@@ -935,12 +881,54 @@ test("project level exports field config", async () => {
     "package.json": JSON.stringify(
       {
         name: "repo",
+        preconstruct: { packages: ["packages/*"], exports: true },
+      },
+      null,
+      2
+    ),
+    "packages/pkg-a/package.json": JSON.stringify({
+      name: "pkg-a",
+    }),
+    "packages/pkg-a/src/index.js": "",
+  });
+  await fix(tmpPath);
+  expect(await getFiles(tmpPath, ["**/package.json"])).toMatchInlineSnapshot(`
+    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ package.json ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
+    {
+      "name": "repo",
+      "preconstruct": {
+        "packages": [
+          "packages/*"
+        ],
+        "exports": true
+      }
+    }
+    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ packages/pkg-a/package.json ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
+    {
+      "name": "pkg-a",
+      "main": "dist/pkg-a.cjs.js",
+      "module": "dist/pkg-a.esm.js",
+      "exports": {
+        ".": {
+          "module": "./dist/pkg-a.esm.js",
+          "default": "./dist/pkg-a.cjs.js"
+        },
+        "./package.json": "./package.json"
+      }
+    }
+
+  `);
+});
+
+test("experimental exports flag is removed", async () => {
+  const tmpPath = await testdir({
+    "package.json": JSON.stringify(
+      {
+        name: "repo",
         preconstruct: {
           packages: ["packages/*"],
           exports: true,
-          ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-            exports: true,
-          },
+          ___experimentalFlags_WILL_CHANGE_IN_PATCH: { exports: true },
         },
       },
       null,
@@ -960,12 +948,10 @@ test("project level exports field config", async () => {
         "packages": [
           "packages/*"
         ],
-        "exports": true,
-        "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-          "exports": true
-        }
+        "exports": true
       }
     }
+
     ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ packages/pkg-a/package.json ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
     {
       "name": "pkg-a",

--- a/packages/cli/src/__tests__/validate.ts
+++ b/packages/cli/src/__tests__/validate.ts
@@ -675,9 +675,6 @@ describe("exports field config", () => {
         },
         preconstruct: {
           exports: config,
-          ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-            exports: true,
-          },
         },
       }),
       "src/index.js": "",
@@ -761,9 +758,6 @@ describe("exports field config", () => {
             module: "dist/pkg-a.esm.js",
             preconstruct: {
               exports: config,
-              ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-                exports: true,
-              },
             },
           }),
           "src/index.js": "",
@@ -781,9 +775,6 @@ test("no module field with exports field", async () => {
       main: "dist/pkg-a.cjs.js",
       preconstruct: {
         exports: true,
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -805,9 +796,6 @@ test("has browser field but no browser condition", async () => {
       },
       preconstruct: {
         exports: true,
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -828,9 +816,6 @@ test("has browser condition but no browser field", async () => {
         exports: {
           envConditions: ["browser"],
         },
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
@@ -848,14 +833,44 @@ test("preconstruct.exports: true no exports field", async () => {
       module: "dist/pkg-a.esm.js",
       preconstruct: {
         exports: true,
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     "src/index.js": "",
   });
   await expect(validate(tmpPath)).rejects.toMatchInlineSnapshot(
     `[Error: exports field was not found, expected \`{".":{"module":"./dist/pkg-a.esm.js","default":"./dist/pkg-a.cjs.js"},"./package.json":"./package.json"}\`]`
+  );
+});
+
+test("experimental exports flag is removed", async () => {
+  const tmpPath = await testdir({
+    "package.json": JSON.stringify(
+      {
+        name: "repo",
+        preconstruct: {
+          packages: ["packages/*"],
+          exports: true,
+          ___experimentalFlags_WILL_CHANGE_IN_PATCH: { exports: true },
+        },
+      },
+      null,
+      2
+    ),
+    "packages/pkg-a/package.json": JSON.stringify({
+      name: "pkg-a",
+      main: "dist/pkg-a.cjs.js",
+      module: "dist/pkg-a.esm.js",
+      exports: {
+        ".": {
+          module: "./dist/pkg-a.esm.js",
+          default: "./dist/pkg-a.cjs.js",
+        },
+        "./package.json": "./package.json",
+      },
+    }),
+    "packages/pkg-a/src/index.js": "",
+  });
+  await expect(validate(tmpPath)).rejects.toMatchInlineSnapshot(
+    `[Error: The behaviour from the experimental flag "exports" is the current behaviour now, the flag should be removed]`
   );
 });

--- a/packages/cli/src/build/__tests__/build.ts
+++ b/packages/cli/src/build/__tests__/build.ts
@@ -684,9 +684,6 @@ test("worker and browser build", async () => {
         exports: {
           envConditions: ["browser", "worker"],
         },
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-          exports: true,
-        },
       },
     }),
     node_modules: {
@@ -741,7 +738,6 @@ test("typescript with nodenext module resolution", async () => {
       name: "@exports/repo",
       preconstruct: {
         packages: ["packages/pkg-a"],
-        ___experimentalFlags_WILL_CHANGE_IN_PATCH: { exports: true },
       },
     }),
     "packages/pkg-a/package.json": JSON.stringify({

--- a/packages/cli/src/package.ts
+++ b/packages/cli/src/package.ts
@@ -323,9 +323,6 @@ export class Package extends Item<{
   }
 
   exportsFieldConfig(): CanonicalExportsFieldConfig {
-    if (!this.project.experimentalFlags.exports) {
-      return;
-    }
     let defaultExportsFieldEnabled = false;
     if (this.project.directory !== this.directory) {
       const exportsFieldConfig = this.project.json.preconstruct.exports;

--- a/packages/cli/src/project.ts
+++ b/packages/cli/src/project.ts
@@ -27,7 +27,6 @@ export class Project extends Item<{
     distFilenameStrategy?: JSONValue;
     exports?: JSONValue;
     ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
-      exports?: JSONValue;
       logCompiledFiles?: JSONValue;
       typeScriptProxyFileWithImportEqualsRequireAndExportEquals?: JSONValue;
       keepDynamicImportAsDynamicImportInCommonJS?: JSONValue;
@@ -38,7 +37,6 @@ export class Project extends Item<{
     let config =
       this.json.preconstruct.___experimentalFlags_WILL_CHANGE_IN_PATCH || {};
     return {
-      exports: !!config.exports,
       logCompiledFiles: !!config.logCompiledFiles,
       typeScriptProxyFileWithImportEqualsRequireAndExportEquals: !!config.typeScriptProxyFileWithImportEqualsRequireAndExportEquals,
       keepDynamicImportAsDynamicImportInCommonJS: !!config.keepDynamicImportAsDynamicImportInCommonJS,

--- a/packages/cli/src/validate.ts
+++ b/packages/cli/src/validate.ts
@@ -125,10 +125,10 @@ export const FORMER_FLAGS_THAT_ARE_ENABLED_NOW = new Set<string>([
   "newEntrypoints",
   "newDistFilenames",
   "newProcessEnvNodeEnvReplacementStrategyAndSkipTerserOnCJSProdBuild",
+  "exports",
 ]);
 
 export const EXPERIMENTAL_FLAGS = new Set([
-  "exports",
   "logCompiledFiles",
   "typeScriptProxyFileWithImportEqualsRequireAndExportEquals",
   "keepDynamicImportAsDynamicImportInCommonJS",

--- a/site/docs/configuration.md
+++ b/site/docs/configuration.md
@@ -132,7 +132,7 @@ Packages map 1:1 with npm packages. Along with specifying the `entrypoints` opti
 }
 ```
 
-### `exports` (experimental) {#exports}
+### `exports` {#exports}
 
 ```ts
 | boolean
@@ -154,21 +154,7 @@ Using the `exports` field enables a couple of things:
 
 Note that adding an `exports` field can arguably be a breaking change, you may want to use the `extra` option to add more exports so that imports that worked previously still work or only add the `exports` field in a major version.
 
-To opt into this experimental feature, you must enable it in the root of your project by setting the `exports` experimental flag in your preconstruct config section of your `package.json` file to `true`.
-
-```diff
-{
-  "name": "@sample/repo",
-  "version": "0.0.0",
-  "preconstruct": {
-+    "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
-+      "exports": true
-+    },
-  }
-}
-```
-
-The `exports` field feature then needs to be enabled, you can do this at the project or package level like this. The `envConditions` and `extra` options can only be configured at a package level.
+The `exports` field feature can be enabled at the project or package level like this. The `envConditions` and `extra` options can only be configured at a package level.
 
 ```diff
 {

